File: bf/bfi1.rs
#![feature(io)]   // to allow `input.chars()`:

use std::env;
use std::fs::File;

use std::io;
use std::io::prelude::*;

use std::num::Wrapping;



const ONE: Wrapping<u8> = Wrapping(1);

const MEM_SIZE: usize = 30000;
const SYNTAX_CHARS: &'static str = "+-<>.,[]";

#[derive(Debug, PartialEq)]
enum Op {
    Inc,
    Dec,
    Left,
    Right,
    Print,
    Read,
    Begin,
    End
}

fn compile(prog: &String) -> Vec<Op> {
    let syntax = SYNTAX_CHARS.to_string();
    prog.chars()
        .filter(|c: &char|
            syntax.contains(*c)
        ).map(|c: char| -> Op {
            if      c == '+' { Op::Inc   }
            else if c == '-' { Op::Dec   }
            else if c == '<' { Op::Left  }
            else if c == '>' { Op::Right }
            else if c == '.' { Op::Print }
            else if c == ',' { Op::Read  }
            else if c == '[' { Op::Begin }
            else if c == ']' { Op::End   }

            // "an unknown symbol" here implies "BF has syntax errors":
            else { panic!("Syntax error!") } 
        }).collect()
}

fn find_jump(prog: &Vec<Op>, cp: usize, whom: Op) -> usize {
    let (me, step) = match whom {
        Op::Begin => (Op::End,   -1isize),
        Op::End   => (Op::Begin, 1isize),
        _ => panic!("Not a loop companion: {:?}", whom)
    };

    let mut nest = 0isize;
    let mut i = cp;

    loop {
        i = (i as isize + step) as usize;
        let op = &prog[i];

        if *op == me {
            nest += 1
        } else if *op == whom {
            if nest > 0 {
                nest -= 1
            } else if nest == 0 {
                return i;
            } else {
                panic!("Unmatched loop: {}", i);
            }
        }
    }
}

fn interpret<In: io::Read, Out: io::Write>(prog: &Vec<Op>, input: In, output: &mut Out) {
    let mut mem = [0u8; MEM_SIZE];
    let mut input = input.chars();

    let mut cp = 0;  // code pointer
    let mut dp = 0;  // data pointer
    loop {
        //println!("Executing {:?}", prog[cp]);
        match prog[cp] {
            Op::Inc    => {
                let c = Wrapping(mem[dp]);
                mem[dp] = (c + ONE).0;
            }
            Op::Dec    => {
                let c = Wrapping(mem[dp]);
                mem[dp] = (c - ONE).0;
            }
            Op::Left   => dp -= 1,
            Op::Right  => dp += 1,
            Op::Read   => {
                let c = input.next();
                let c = c.expect("end of stream");
                let c = c.unwrap();
                mem[dp] = c as u8;
            }
            Op::Print  => {
                let c = mem[dp];
                let buf = [c; 1];
                output.write(&buf).unwrap();
            }
            Op::Begin  => {
                let c = mem[dp];
                if c == 0 {
                    cp = find_jump(prog, cp, Op::End);
                    continue;
                }
            }
            Op::End    => {
                let c = mem[dp];
                if c != 0 {
                    cp = find_jump(prog, cp, Op::Begin);
                    continue;
                }
            }
        }
        cp += 1;

        if cp >= prog.len() {
            break;
        }
    } 
}

fn run(source: &String) {
    let ops = compile(source);

    let stdin = io::stdin();
    let mut stdout = io::stdout();

    interpret(&ops, stdin, &mut stdout);
}

fn main() {
    let args: Vec<String> = env::args().collect();

    let progfile = &args[1];

    let mut f = File::open(progfile).expect("file not found");

    let mut contents = String::new();
    f.read_to_string(&mut contents).expect("i/o error");

    run(&contents);
}

File: bf/bfi1.cpp
#include <vector>
#include <string>
#include <fstream>
#include <iostream>
#include <streambuf>
#include <cstdint>


typedef uint32_t addr_t;

const int MEMSIZE = 30000;

const char *syntax = "+-<>.,[]";

enum Op {
    Inc,
    Dec,
    Left,
    Right,
    Print,
    Read,
    Begin,
    End
};

typedef std::vector<Op> Ops;

Ops compile(const std::string &prog) {
    Ops ops;
    for (auto c : prog) {
        switch (c) {
            case '+': ops.push_back(Op::Inc); break;
            case '-': ops.push_back(Op::Dec); break;
            case '<': ops.push_back(Op::Left); break;
            case '>': ops.push_back(Op::Right); break;
            case '.': ops.push_back(Op::Print); break;
            case ',': ops.push_back(Op::Read); break;
            case '[': ops.push_back(Op::Begin); break;
            case ']': ops.push_back(Op::End); break;
        }
    }
    return ops;
}

addr_t find_jump(const Ops& ops, addr_t cp, Op to) {
    int nest = 0;
    addr_t i = cp;

    off_t step;
    Op from;
    switch (to) {
    case Op::Begin:
        from = Op::End;
        step = -1;
        break;
    case Op::End:
        from = Op::Begin;
        step = 1;
        break;
    default:
        throw "find_jump: Op::Begin or Op::End expected";
    }
//    std::cerr << "looking for " << syntax[to] << std::endl;
    while (true) {
        i += step;

        Op op = ops.at(i);
//        std::cerr << "i=" << i << std::endl;

        if (op == from) {
            ++nest;
        } else if (op == to) {
            if (nest > 0) {
                --nest;
            } else if (nest == 0) {
                return i;
            } else {
                throw "find_jump: malformed loop";
            }
        }
    }
}

void interpret(const Ops& ops, std::istream &input, std::ostream &output) {
    uint32_t cp = 0;
    addr_t dp = 0;

    uint8_t mem[MEMSIZE] = { 0 };

    while (true) {
        char c;
        Op op = ops.at(cp);
//        std::cerr << "exe[" << cp << "]: " << syntax[op] << std::endl;

        switch (op) {
        case Op::Inc:
            mem[dp] += 1;
            break;
        case Op::Dec:
            mem[dp] -= 1;
            break;
        case Op::Left:
            dp -= 1;
            break;
        case Op::Right:
            dp += 1;
            break;
        case Op::Print:
            output.put((char)mem[dp]);
            break;
        case Op::Read:
            input >> c;
            mem[dp] = (uint8_t)c;
            break;
        case Op::Begin:
            if (mem[dp] == 0) {
                cp = find_jump(ops, cp, Op::End);
                continue;
            }
            break;
        case Op::End:
            if (mem[dp] != 0) {
                cp = find_jump(ops, cp, Op::Begin);
                continue;
            }
            break;
        }
        ++cp;
        if (cp >= ops.size())
            break;
    }
}

void run(const std::string& source) {
    const Ops ops = compile(source); 

    interpret(ops, std::cin, std::cout);
}

int main(int argc, char **argv) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <source.bf>" << std::endl;
        return 1;
    }
    const char* progfile = argv[1];    

    std::ifstream f(progfile);

    std::string contents = std::string(
        std::istreambuf_iterator<char>(f),
        std::istreambuf_iterator<char>()
    );

    run(contents);
}
